File: vignettes/12_telemetry/csharp/12_telemetry.cs
// Telemetry — Observability and tracing (C#)
//
// This sample demonstrates how to add observability to agents using
// Activity/DiagnosticSource patterns. It mirrors the Julia vignette
// 12_telemetry.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using System.Diagnostics;
using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// ── 1. Activity/Span Basics ─────────────────────────────────────────────────

Console.WriteLine("=== Span Basics (using System.Diagnostics.Activity) ===");

var source = new ActivitySource("AgentFramework.Sample");

using (var activity = source.StartActivity("my_operation", ActivityKind.Client))
{
    activity?.SetTag("custom.key", "some_value");
    activity?.AddEvent(new ActivityEvent("checkpoint_reached",
        tags: new ActivityTagsCollection { { "step", 1 } }));
    activity?.AddEvent(new ActivityEvent("data_processed",
        tags: new ActivityTagsCollection { { "records", 42 } }));
    await Task.Delay(10);
    activity?.SetStatus(ActivityStatusCode.Ok);
    Console.WriteLine($"  Activity: {activity?.OperationName}, "
        + $"duration: {activity?.Duration.TotalMilliseconds:F0}ms, "
        + $"status: {activity?.Status}");
}

// ── 2. In-Memory Span Collection ────────────────────────────────────────────

Console.WriteLine("\n=== In-Memory Span Collection ===");

var collectedSpans = new List<TelemetryRecord>();

// Record spans manually (mirrors InMemoryTelemetryBackend)
var testSpan = new TelemetryRecord("test.operation", "client", DateTimeOffset.UtcNow);
testSpan.Attributes["gen_ai.request.model"] = "qwen3:8b";
testSpan.Finish("ok");
collectedSpans.Add(testSpan);

Console.WriteLine($"  Recorded spans: {collectedSpans.Count}");
Console.WriteLine($"  First span: {collectedSpans[0].Name}");
collectedSpans.Clear();
Console.WriteLine($"  After clear: {collectedSpans.Count}");

// ── 3. Instrumented Workflow with OpenTelemetry ─────────────────────────────

Console.WriteLine("\n=== Instrumented Workflow ===");

var upperExecutor = new UppercaseExecutor();
var reverseExecutor = new ReverseTextExecutor();

var workflow = new WorkflowBuilder(upperExecutor)
    .AddEdge(upperExecutor, reverseExecutor)
    .WithOpenTelemetry(
        configure: cfg => cfg.EnableSensitiveData = true,
        activitySource: source)
    .Build();

await using (Run run = await InProcessExecution.RunAsync(workflow, "Hello, World!"))
{
    foreach (WorkflowEvent evt in run.NewEvents)
    {
        if (evt is ExecutorCompletedEvent executorComplete)
        {
            Console.WriteLine($"  {executorComplete.ExecutorId}: {executorComplete.Data}");
        }
    }
}

// ── 4. Instrumented Agent Call ──────────────────────────────────────────────

Console.WriteLine("\n=== Instrumented Agent ===");

AIAgent agent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        name: "TracedAgent",
        instructions: "You are a helpful assistant. Keep answers brief.");

var agentSpan = new TelemetryRecord("agent.run", "internal", DateTimeOffset.UtcNow);
agentSpan.Attributes["gen_ai.agent.name"] = "TracedAgent";

var chatSpan = new TelemetryRecord("chat.completion", "client", DateTimeOffset.UtcNow);
chatSpan.Attributes["gen_ai.request.model"] = modelName;

var agentResponse = await agent.RunAsync("What is 2 + 2?");
Console.WriteLine($"  Answer: {agentResponse.Text}");

chatSpan.Finish("ok");
agentSpan.Attributes["gen_ai.response.model"] = modelName;
agentSpan.Finish("ok");

collectedSpans.AddRange([agentSpan, chatSpan]);

Console.WriteLine($"\n  Total spans: {collectedSpans.Count}");
foreach (var s in collectedSpans)
{
    var model = s.Attributes.GetValueOrDefault("gen_ai.request.model", "—");
    Console.WriteLine($"    [{s.Kind}] {s.Name} — {s.DurationMs}ms, "
        + $"model={model}, status={s.Status}");
}

// ── Supporting Types ────────────────────────────────────────────────────────

/// <summary>Simple telemetry record mirroring TelemetrySpan.</summary>
class TelemetryRecord
{
    public string Name { get; }
    public string Kind { get; }
    public DateTimeOffset StartTime { get; }
    public DateTimeOffset? EndTime { get; private set; }
    public string Status { get; private set; } = "unset";
    public Dictionary<string, object> Attributes { get; } = new();

    public long? DurationMs => EndTime.HasValue
        ? (long)(EndTime.Value - StartTime).TotalMilliseconds : null;

    public TelemetryRecord(string name, string kind, DateTimeOffset startTime)
    {
        Name = name;
        Kind = kind;
        StartTime = startTime;
    }

    public void Finish(string status = "ok")
    {
        EndTime = DateTimeOffset.UtcNow;
        Status = status;
    }
}

/// <summary>Converts input text to uppercase.</summary>
internal sealed class UppercaseExecutor() : Executor<string, string>("UppercaseExecutor")
{
    public override async ValueTask<string> HandleAsync(string message,
        IWorkflowContext context, CancellationToken cancellationToken = default)
        => message.ToUpperInvariant();
}

/// <summary>Reverses input text.</summary>
internal sealed class ReverseTextExecutor() : Executor<string, string>("ReverseTextExecutor")
{
    public override async ValueTask<string> HandleAsync(string message,
        IWorkflowContext context, CancellationToken cancellationToken = default)
        => new(message.Reverse().ToArray());
}

File: vignettes/12_telemetry/python/12_telemetry.py
"""
Telemetry — Observability and tracing (Python)

This sample demonstrates how to add observability to agents using
OpenTelemetry-style tracing patterns. It mirrors the Julia vignette
12_telemetry.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from agent_framework.ollama import OllamaChatClient


# ── Telemetry Span ───────────────────────────────────────────────────────────


@dataclass
class TelemetrySpan:
    """Represents a span/activity in a trace, following OpenTelemetry semantics."""

    name: str
    kind: str = "internal"
    start_time: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    end_time: datetime | None = None
    status: str = "unset"
    attributes: dict[str, Any] = field(default_factory=dict)
    events: list[dict[str, Any]] = field(default_factory=list)

    def add_event(self, name: str, attributes: dict[str, Any] | None = None) -> None:
        """Append a timestamped event to the span."""
        self.events.append({
            "name": name,
            "time": datetime.now(timezone.utc),
            "attributes": attributes or {},
        })

    def finish(self, status: str = "ok") -> None:
        """Mark the span as finished."""
        self.end_time = datetime.now(timezone.utc)
        self.status = status

    @property
    def duration_ms(self) -> int | None:
        """Span duration in milliseconds."""
        if self.end_time is None:
            return None
        delta = self.end_time - self.start_time
        return int(delta.total_seconds() * 1000)


# ── GenAI Semantic Conventions ───────────────────────────────────────────────

# Following OpenTelemetry Semantic Conventions for Generative AI
GEN_AI_REQUEST_MODEL = "gen_ai.request.model"
GEN_AI_RESPONSE_MODEL = "gen_ai.response.model"
GEN_AI_USAGE_INPUT_TOKENS = "gen_ai.usage.input_tokens"
GEN_AI_USAGE_OUTPUT_TOKENS = "gen_ai.usage.output_tokens"
GEN_AI_AGENT_NAME = "gen_ai.agent.name"
GEN_AI_TOOL_NAME = "gen_ai.tool.name"


# ── In-Memory Telemetry Backend ──────────────────────────────────────────────


class InMemoryTelemetryBackend:
    """Stores spans in memory for testing and debugging."""

    def __init__(self) -> None:
        self.spans: list[TelemetrySpan] = []

    def record_span(self, span: TelemetrySpan) -> None:
        self.spans.append(span)

    def get_spans(self) -> list[TelemetrySpan]:
        return list(self.spans)

    def clear(self) -> None:
        self.spans.clear()


# ── Instrumented Agent Wrapper ───────────────────────────────────────────────


class InstrumentedAgent:
    """Wraps an agent to emit telemetry spans on each call."""

    def __init__(self, agent: Any, backend: InMemoryTelemetryBackend, name: str = "agent"):
        self.agent = agent
        self.backend = backend
        self.name = name

    async def run(self, prompt: str) -> str:
        """Run the agent with telemetry instrumentation."""
        # Agent-level span
        agent_span = TelemetrySpan(name="agent.run", kind="internal")
        agent_span.attributes[GEN_AI_AGENT_NAME] = self.name
        agent_span.attributes["message_count"] = 1

        # Chat-level span
        chat_span = TelemetrySpan(name="chat.completion", kind="client")
        chat_span.attributes[GEN_AI_REQUEST_MODEL] = "qwen3:8b"

        try:
            result = await self.agent.run(prompt)
            result_text = str(result)

            chat_span.finish(status="ok")
            self.backend.record_span(chat_span)

            agent_span.attributes[GEN_AI_RESPONSE_MODEL] = "qwen3:8b"
            agent_span.finish(status="ok")
            self.backend.record_span(agent_span)

            return result_text

        except Exception as e:
            chat_span.add_event("exception", {"type": type(e).__name__, "message": str(e)})
            chat_span.finish(status="error")
            self.backend.record_span(chat_span)

            agent_span.add_event("exception", {"type": type(e).__name__, "message": str(e)})
            agent_span.finish(status="error")
            self.backend.record_span(agent_span)
            raise


async def main() -> None:
    # ── 1. Span Basics ───────────────────────────────────────────────────
    print("=== Span Basics ===")
    span = TelemetrySpan(name="my_operation", kind="client")
    span.attributes["custom.key"] = "some_value"
    span.add_event("checkpoint_reached", {"step": 1})
    span.add_event("data_processed", {"records": 42})
    time.sleep(0.01)
    span.finish(status="ok")
    print(f"  Span: {span.name}, duration: {span.duration_ms}ms, "
          f"events: {len(span.events)}, status: {span.status}")

    # ── 2. In-Memory Backend ─────────────────────────────────────────────
    print("\n=== In-Memory Backend ===")
    backend = InMemoryTelemetryBackend()
    backend.record_span(span)
    print(f"  Recorded spans: {len(backend.get_spans())}")
    backend.clear()
    print(f"  After clear: {len(backend.get_spans())}")

    # ── 3. Instrumented Agent ────────────────────────────────────────────
    print("\n=== Instrumented Agent ===")
    client = OllamaChatClient(host="http://localhost:11434", model_id="qwen3:8b")
    agent = client.as_agent(
        name="TracedAgent",
        instructions="You are a helpful assistant. Keep answers brief.",
    )

    backend = InMemoryTelemetryBackend()
    traced = InstrumentedAgent(agent, backend, name="TracedAgent")

    result = await traced.run("What is 2 + 2?")
    print(f"  Answer: {result}")

    print(f"\n  Total spans: {len(backend.get_spans())}")
    for s in backend.get_spans():
        model = s.attributes.get(GEN_AI_REQUEST_MODEL, "—")
        print(f"    [{s.kind}] {s.name} — {s.duration_ms}ms, "
              f"model={model}, status={s.status}")


if __name__ == "__main__":
    asyncio.run(main())
